Why Do LLMs Design Mediocre Architecture? (#4)
Recurse-ML/recurse-ml.github.io: content/arch.md

+++
date = '2025-09-14T15:45:21-07:00'
draft = true
title = 'Why Do LLMs Design Mediocre Architecture?'
+++

LLMs create mediocre architecture that compounds technical debt.
By "mediocre architecture" I mean code that's functionally correct but hard to maintain and expand.
In our example it's duplicated code without opinionated abstractions.

Where should we expect them to fail?

> How can tech leads benefit from increases in velocity without eventually crumbling under tech debt?

Reviewer: ```suggestion
How can tech leads to benefits from increases in velocity without eventually crumbling under tech debt?
```

Reviewer 2: I think the original grammar was correct?

How should teams change their working style as a result?

## Worked Example: Implementing Email Notifications

> I ran Claude Code in the [FastAPI Example App](http://github.com/fastapi/fastapi-example-app) with the following prompt:

Reviewer: Link is 404

                )
                send_email(
                    email_to=db_user.email,
                    subject=email_data.subject,
                    html_content=email_data.html_content,
                )
            except Exception as e:
                logger.info(f"Failed to send admin update email to {db_user.email}: {e}")

    return db_user
```

Here Claude has successfully understood the existing code structure and has created a new function that follows a similar pattern as before.
The functions to generate emails and their usage are nearly identical with the existing code.

But here's the thing, if this were a PR, I'd reject it.

### Why This Doesn't Work

The existing approach to sending notification emails has reached its critical mass.
We now have 6 duplicate functions doing essentially the same thing.
About 60% of the endpoint code is just handling notification emails.

There are also two dangerous gotchas in the code:

1. The code catches bare `Exception`s and logs them at `info` level - a surefire way to leave logical bugs undiscovered forever.
2. The HTML templates used by generator methods don't actually exist.

### Conflicting Optimization Objectives

> Better design would require removing the existing email function and replacing it with an abstraction.

Reviewer 2: ```suggestion
A better design would be to remove the existing email function and replace it with an abstraction.
```

But that means making opinionated decisions about the codebase.

This is RLHF-induced sycophancy at work.

> Not only does it respond to you with "you're perfectly right!", it considers your code as such.

Reviewer 2: ```suggestion
Not only does it respond to you with "you're perfectly right!", it considers your code equally as "perfectly right".
```

Well, unless you're terribly off course.
Claude saw one way of doing emails and assumed that was the way.

From a product perspective, this makes sense. I rarely want an LLM coding agent making radical changes to my codebase.

Instead of being a problem, this is an opportunity to delineate responsibilities.

> ## Reclaim Your Role as Architect

Reviewer 2: ```suggestion
## Reclaim Your Role as an Architect
```

Reviewer 2: idk if an is correct here maybe `The Architect`? `as Architect` just sounds a bit funny to me

When LLMs handle the implementation details, our role as reviewers fundamentally changes.
We're no longer line-by-line code inspectors catching syntax errors or style violations.
Instead, we become architectural guardians, focusing on the bigger questions:
Are we solving the right problem?
Is this the right abstraction?
Will this scale?

> This shift isn't just about adapting to AI tools; it's about reclaiming the most intellectually rewarding part of software development—the design decisions that shape how our systems grow and evolve.

Reviewer 2: idk how picky we want to be about the blog post appearing like it has been written by an LLM but if we are then it's worth flagging that the "This isn't about X; it's about Y" pattern is one that's overly used by LLMs

Reviewer 2: same goes to the em dash :P

The first line of defence is a good ol' whiteboard meeting.
I've noticed myself pulling a colleague aside for a design discussion more frequently.
Especially as I know that once I have a solid high-level design, the (LLM-assisted) implementation will be easy.

Sometimes, just implementing a feature is faster than having a long discussion around it.
Normally, I'm hesitant to suggest drastic architectural changes in PRs.
The tradeoff between clean software architecture and time taken to implement it is rarely worth it in a startup.

> Especially, if there's an existing solution.

Reviewer: ```suggestion
Especially if there's an existing solution.
```

However, as the code can be easily generated, that's no longer a valid concern.
That's why, when reviewing PRs, I now pay more attention to high-level design decisions.

As a reviewer, it is difficult to think about both the big picture design and line-by-line implementation all at once.
That's why I outsource most of the low-level thinking to automated tooling.
Tests and classic standard analyzers are helpful here.
Obviously we use Recurse ML as our safety net.

However, one issue I find myself battling with is that LLMs tend to repeat the same mistakes over and over again.
This is where [custom rules](https://docs.recurse.ml/gh/configs/rules/) are particularly helpful.
Think of them as code review comments that automatically check every PR.
When the tool spots code that violates these patterns, it leaves a comment explaining the issue.

> I've learned from [Nate](https://natesesti.com) @ Continue.dev the following rule of thumb:

Reviewer 2: Do we want to href continue.dev?

> Whenever I leave a comment in a PR, I ask myself, should this be an automatic rule?
> Most of the time the answer is "yes" and I create it.

For example, I created [`bare_exceptions.md`](https://github.com/continuedev/awesome-rules/blob/main/rules/recurse-ml/bare_exceptions.md) to prevent the bare `Exception` case from making it into our production codebase.

## Conclusion

LLMs create mediocre architecture because they optimize for consistency with existing patterns, not optimal design.

> This isn't a bug—it's how they're trained to be helpful rather than opinionated.

Reviewer 2: Another example of the "This isn't X-it's Y". Depends on how ok we are with this appearing as AI generated. Very small nitpicks- sorry hehe

This limitation is an opportunity.
When LLMs handle implementation, we get to focus on the hard architectural decisions that actually matter.

The practical takeaways:
1. Design first, implement second. Pull a colleague aside for that whiteboard session.

> 2. Review PRs for architecture, not syntax. Let tests and linters catch the small stuff.

Reviewer 2: We should mention Recurse ML here, no? Something like
`2. Review PRs for architecture, not syntax. Let tools like Recurse ML (href), tests, and linters catch the small stuff.`

> 3. Turn repeated review comments into automated rules. If you're saying it twice, automate it.

Reviewer 2: Again, maybe we should name drop Recurse ML or link to it. Backlinks and mentioning is good for SEO

4. Don't hesitate to suggest refactoring. When implementation is cheap, good design becomes affordable.

Embrace the promise of LLMs to free humans from rote work towards creative high-level tasks.
Designing elegant software is hard.
It requires understanding existing code, having a mental model of people who will maintain it and the environment it will run in.
However, it's also fun.
It's creative in the same way that writing or visual art is creative.
The space of viable solutions is vast.
Greatness is highly subjective and context-dependent.

> But creating something, that other people appreciate and build upon is one of the most rewarding parts of the human experience.

Reviewer: ```suggestion
But creating something that other people appreciate and build upon is one of the most rewarding parts of the human experience.
```
